napari_animation/animation.py

```diff
@@ -28,15 +29,21 @@ class Animation:
         Dictionary relating state attributes to interpolation functions.
     """
 
+    key_frames: EventedList = Field(
+        default_factory=EventedList, allow_mutation=False
+    )
+    frame: int = -1
+
     def __init__(self, viewer):
+        # allow extra attributes during model initialization, useful for mixins
+        self.__config__.extra = Extra.allow
+        super().__init__()
         self.viewer = viewer
```

Comment:
Thanks a lot for the info ! 😄 for the `state_interpolation_map` it is true that for now it's kinda constant, but we thought it might be good at some point to allow the user to play with custom functions. 

Also for the `viewer`, I added `config.extra` mostly because I thought it would allow not to get an `animation.events.viewer`, as we probably wouldn't need it. Am I wrong here ? Is there an other way to get the same thing, or is it not that important anyways ?

Thanks again ! 👍

Reply:
> but we thought it might be good at some point to allow the user to play with custom functions.

makes sense... maybe it would make for a good `allow_mutation=False` field (since you don't need the dict itself to change... just its members)

> it would allow not to get an `animation.events.viewer`, as we probably wouldn't need it.

gotcha... [here](https://github.com/napari/napari/blob/336772a7d4123795dccb644d75adbb203efae360/napari/utils/events/evented_model.py#L131-L137) are the key lines to know about in `EventedModel.__init__`:
```python
# add event emitters for each field which is mutable
fields = []
for name, field in self.__fields__.items():
    if field.field_info.allow_mutation:
        fields.append(name)
```

So, there are few ways to hide an attribute from events:  `_underscore` attributes, `ClassVars`, and `pydantic.Field(allow_mutation=False)`.

example:

```python
from napari.utils.events import EventedModel
from pydantic import Field
from typing import ClassVar

class M(EventedModel):
    w: int = 1
    _x: int = 1
    y: ClassVar[int] = 1
    z: int = Field(default=1, allow_mutation=False)
```

```python
In [47]: M.__fields__
Out[47]:
{'w': ModelField(name='w', type=int, required=False, default=1),
 'z': ModelField(name='z', type=int, required=False, default=1)}
# note that '_x' and 'y' are not model "fields"
# and therefore won't get events

In [48]: M().events.emitters
Out[48]: {'w': <napari.utils.events.event.EventEmitter at 0x16c1b15e0>}
# note that "z" has no event
```

Reply:
Such a good resource ! Thanks for this, it helped a lot :) I think I got it now, and I'm ready to propose something ; I can switch `viewer`to a private `_viewer` if everyone thinks we're ready for it ! Otherwise I'll reverse to a allow_mutation=false.
Thanks again !!